Fix segmentation fault

jruot opened the pull request:
Resizing window in hello_triangle sample caused segmentation fault

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/KhronosGroup/Vulkan-Samples?pullRequest=157) <br/>All committers have signed the CLA.

wasimabbas-arm commented on samples/api/hello_triangle/hello_triangle.cpp:
```diff
@@ -1096,6 +1096,7 @@ void HelloTriangle::update(float delta_time)
 	if (res == VK_SUBOPTIMAL_KHR || res == VK_ERROR_OUT_OF_DATE_KHR)
 	{
 		resize(context.swapchain_dimensions.width, context.swapchain_dimensions.height);
+		return;
```
Where exactly are you getting a getting a segmentation fault? This doesn't seem like the right solution. You are basically skipping a frame. You would want to render to the new dimensions right?

JoseEmilio-ARM replied:
Agree, please provide platform details and stack trace, I cannot reproduce on Windows. Would it be possible to handle this in `resize`?

jruot replied:
Thanks - I just started learning vulkan so please bear with me.
Platform: arch linux 5.7.9-zen1-1-zen
GPU: AMD RADV NAVI10 (LLVM 10.0.0)

I'm getting weird `index` from the `acquire_next_image` function after resizing the window and it goes over the array bounds on the line inside `render_triangle` function:
`VkCommandBuffer cmd = context.per_frame[swapchain_index].primary_command_buffer;`
After this the program crashes.
The `vkAcquireNextImageKHR` function returns `VK_ERROR_OUT_OF_DATE_KHR`. Should the function get the next image index even when the result is `VK_ERROR_...`?

jruot replied:
I investigated little bit more and found this https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#vkAcquireNextImageKHR
> If the swapchain images no longer match native surface properties, either VK_SUBOPTIMAL_KHR or VK_ERROR_OUT_OF_DATE_KHR must be returned. If VK_ERROR_OUT_OF_DATE_KHR is returned, no image is acquired and attempts to present previously acquired images to the swapchain will also fail with VK_ERROR_OUT_OF_DATE_KHR. 

In that case would better solution be something like this:
```
	if (res == VK_SUBOPTIMAL_KHR || res == VK_ERROR_OUT_OF_DATE_KHR)
	{
		resize(context.swapchain_dimensions.width, context.swapchain_dimensions.height);
                if (res == VK_ERROR_OUT_OF_DATE_KHR)
                {
                     return;
                }
	}
```

ZandroFargnoli replied:
@Kuraturpa could you confirm for me if the segmentation fault disappears when you change that section of code with the following code snippet:

```
uint32_t index;

auto res = acquire_next_image(context, &index);

// Handle outdated error in acquire.
if (res == VK_SUBOPTIMAL_KHR || res == VK_ERROR_OUT_OF_DATE_KHR)
{
	resize(context.swapchain_dimensions.width, context.swapchain_dimensions.height);
	res = acquire_next_image(context, &index);
}

if (res != VK_SUCCESS)
{
	vkQueueWaitIdle(context.queue);
	return;
}
```

We recreate the swapchain in `resize()`, so we will then need to acquire an image again. If this still fails then we can skip the frame with the return.

jruot replied:
@ZandroFargnoli This fixes it. The second `acquire_next_image` call returns `VK_SUCCESS`.

jruot committed:
Fix segmentation fault in hello_triangle

Returned image index was invalid after resizing the sample window.

Acquire image again after recreating the swapchain.

ZandroFargnoli reviewed (approved):
LGTM!
